internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -721,14 +731,218 @@ func (r *Reconciler) deleteMachineDeployment(ctx context.Context, cluster *clust
+	r.recorder.Eventf(cluster, corev1.EventTypeNormal, createEventReason, "Created %q", tlog.KObj{Obj: mp.Object})
+
+	// Wait until MachinePool is visible in the cache.
+	// Note: We have to do this because otherwise using a cached client in current state could
+	// miss a newly created MachinePool (because the cache might be stale).
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, 5*time.Second, true, func(ctx context.Context) (bool, error) {
+		key := client.ObjectKey{Namespace: mp.Object.Namespace, Name: mp.Object.Name}
+		if err := r.Client.Get(ctx, key, &expv1.MachinePool{}); err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed waiting for MachinePool %s to be visible in the cache after create", mp.Object.Kind)
+	}
+
+	return nil
+}
+
+// updateMachinePool updates a MachinePool. Also rotates the corresponding Templates if necessary.
+func (r *Reconciler) updateMachinePool(ctx context.Context, cluster *clusterv1.Cluster, currentMP, desiredMP *scope.MachinePoolState) error {
+	log := tlog.LoggerFrom(ctx).WithMachinePool(desiredMP.Object)
+
+	infraCtx, _ := log.WithObject(desiredMP.InfrastructureMachinePoolObject).Into(ctx)
+	if err := r.reconcileReferencedObject(infraCtx, reconcileReferencedObjectInput{
+		cluster:       cluster,
+		current:       currentMP.InfrastructureMachinePoolObject,
+		desired:       desiredMP.InfrastructureMachinePoolObject,
+		versionGetter: contract.ControlPlane().Version().Get,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to reconcile %s", tlog.KObj{Obj: currentMP.Object})
+	}
+
+	bootstrapCtx, _ := log.WithObject(desiredMP.BootstrapObject).Into(ctx)
+	if err := r.reconcileReferencedObject(bootstrapCtx, reconcileReferencedObjectInput{
+		cluster:       cluster,
+		current:       currentMP.BootstrapObject,
+		desired:       desiredMP.BootstrapObject,
+		versionGetter: contract.ControlPlane().Version().Get,
+	}); err != nil {
+		return errors.Wrapf(err, "failed to reconcile %s", tlog.KObj{Obj: currentMP.Object})
+	}
+
+	// Check differences between current and desired MachinePool, and eventually patch the current object.
+	log = log.WithObject(desiredMP.Object)
+	patchHelper, err := r.patchHelperFactory(ctx, currentMP.Object, desiredMP.Object)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: currentMP.Object})
+	}
+	if !patchHelper.HasChanges() {
+		log.V(3).Infof("No changes for %s", tlog.KObj{Obj: currentMP.Object})
+		return nil
+	}
+
+	log.Infof("Patching %s", tlog.KObj{Obj: currentMP.Object})
+	if err := patchHelper.Patch(ctx); err != nil {
+		return errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: currentMP.Object})
+	}
+	r.recorder.Eventf(cluster, corev1.EventTypeNormal, updateEventReason, "Updated %q%s", tlog.KObj{Obj: currentMP.Object}, logMachinePoolVersionChange(currentMP.Object, desiredMP.Object))
+
```

Comment:
We are missing the polling for the resourceversion change here.

```go
	// Wait until MachinePool is updated in the cache.
	// Note: We have to do this because otherwise using a cached client in current state could
	// return a stale state of a MachinePool we just patched (because the cache might be stale).
	// Note: It is good enough to check that the resource version changed. Other controllers might have updated the
	// MachinePool as well, but the combination of the patch call above without a conflict and a changed resource
	// version here guarantees that we see the changes of our own update.
	err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, 5*time.Second, true, func(ctx context.Context) (bool, error) {
		key := client.ObjectKey{Namespace: currentMP.Object.GetNamespace(), Name: currentMP.Object.GetName()}
		cachedMP := &clusterv1.MachinePool{}
		if err := r.Client.Get(ctx, key, cachedMP); err != nil {
			return false, err
		}
		return currentMP.Object.GetResourceVersion() != cachedMP.GetResourceVersion(), nil
	})
	if err != nil {
		return errors.Wrapf(err, "failed waiting for MachinePool %s to be updated in the cache after patch", tlog.KObj{Obj: currentMP.Object})
	}
```